packages/casting/src/follower-cosmjs.js

```diff
@@ -112,290 +145,442 @@ export const makeCosmjsFollower = (
+      let rightBlockHeight = currentBlockHeight;
+      const rightStreamCellAndData = await getStreamCellAtHeight(
+        rightBlockHeight,
+      );
+      if (!rightStreamCellAndData) {
+        await E(leader).jitter(where);
+        continue;
       }
-      for (let i = 0; i < streamCell.values.length; i += 1) {
-        const data = JSON.parse(streamCell.values[i]);
-        const isLast = i + 1 === streamCell.values.length;
-        const value = /** @type {T} */ (
-          unserializer
-            ? // eslint-disable-next-line no-await-in-loop,@jessie.js/no-nested-await
-              await E(unserializer).unserialize(data)
-            : data
+      let { data: rightData, streamCell: rightStreamCell } =
+        rightStreamCellAndData;
+      const currentData = rightData;
+      const cells = [];
+      // Compare block cell data pairwise (left, right).
+      while (rightBlockHeight > cursorBlockHeight) {
+        // Stream cells should have a numeric block height that is between
+        // 1 (the first valid block height) and the height of the block
+        // they were published in, but we clamp the figure into the valid
+        // range just in case.
+        const leftBlockHeight = clamp(
+          1,
+          rightBlockHeight - 1,
+          rightStreamCell.blockHeight - 1,
         );
-        // QUESTION: How would reach a point where this `isValid()` fails,
-        // and what is the proper handling?
-        if (!unserializer || committer.isValid()) {
-          committer.commit({ value });
-          if (!isLast) {
-            committer = prepareUpdateInOrder();
-          }
+        // Do not scan behind the cusor.
+        if (leftBlockHeight <= cursorBlockHeight) {
+          break;
+        }
+        const leftStreamCellAndData = await getStreamCellAtHeight(
+          leftBlockHeight,
+        );
+        // Do not scan behind a cell with no data.
+        // This should not happen but can be tolerated.
+        if (!leftStreamCellAndData) {
+          break;
         }
+        const { data: leftData, streamCell: leftStreamCell } =
+          leftStreamCellAndData;
+        // Stream cells include a block height that is guaranteed to change
+        // between iterations even if the values are identical.
+        // We can rely on this difference to ensure that we yield
+        // every iteration, including duplicates.
+        // Legacy cells do not contain a block height to distingish versions,
+        // so we simply assume that the value must change between iterations
+        // for a cell to be worthy of notice.
+        if (!arrayEqual(leftData, rightData)) {
+          cells.push(rightStreamCell);
+        }
+        // Prepare for next iteration by moving left to right.
+        rightData = leftData;
+        rightStreamCell = leftStreamCell;
+        rightBlockHeight = leftBlockHeight;
       }
-    };
-
-    const changeFollower = E(leader).watchCasting(castingSpecP);
```

Comment:
FFR, we will need to use something like the `changeFollower` to ensure our RPC servers don't collapse under load.  The `jitter` mechanism is intended just to provide a little bit of randomness, not to wait for the next block to appear.

Reply:
I’ve created an issue to track this https://github.com/Agoric/agoric-sdk/issues/6154

I assume we need to use a Tendermint API that long-polls for block height changes or more specifically for changes to a specific vstorage cell. If you’ve got hints, please post them there.